File: Programmers/LV2/17677.py
# 뉴스 클러스터링


def solution(str1, str2):
    
    answer = 0
    # 입력값 대문자 변환
    str1 = str1.upper()
    str2 = str2.upper()
    # 같은 값일때
    if str1 == str2:
        return 65536
    # 입력값 딕셔너리
    str1Li= {}
    str2Li= {}
    # 교집합 카운트
    count = 0
    # 경우의 수 추출
    for i in range(len(str1)-1):
        tmp = str1[i]+str1[i+1]
        # 문자열로 구성되어 있을때만 넣어줌
        if tmp.isalpha():
            if str1[i]+str1[i+1] not in str1Li:
                str1Li[str1[i]+str1[i+1]] = 1
            else:
                str1Li[str1[i]+str1[i+1]] += 1 
                
    for i in range(len(str2)-1):
        tmp = str2[i]+str2[i+1]
        # 문자열로 구성되어 있을때만 넣어줌
        if tmp.isalpha():
            if str2[i]+str2[i+1] not in str2Li:
                str2Li[str2[i]+str2[i+1]] = 1
            else:
                str2Li[str2[i]+str2[i+1]] += 1 
    
    # 교집합 찾기
    # setLi = list(set(str1Li) & set(str2Li))
    for str1s in str1Li:
        if str1s in str2Li:
            count += min(str1Li[str1s],str2Li[str1s])
            
    # 합집합 찾기
    for str1s in str1Li:
        if str1s in str2Li:
            if str2Li[str1s] < str1Li[str1s]:
                str2Li[str1s] = str1Li[str1s]
        # 반대집합에 없을때 추가
        else:
            str2Li[str1s] = str1Li[str1s]
            
    if count == 0 or sum(str2Li.values()) == 0:
        answer = 65536
    else:
        answer = int((count/ sum(str2Li.values()) ) * 65536)
    return answer


solution('E=M*C^2', 'e=m*c^2')

File: Programmers/LV2/17677.js
// 17677;
// 뉴스 클러스터링
function solution(str1, str2) {
  // 예외 처리(대소문자)
  str1 = str1.toUpperCase();
  str2 = str2.toUpperCase();
  let str1Dic = {};
  let str2Dic = {};
  // 1. 원소 구하기
  for (let i = 0; i < str1.length - 1; i++) {
    let curWord = str1[i] + str1[i + 1];
    curWord = curWord.match(/[a-zA-Z]+/g);
    if (curWord === null || curWord[0].length === 1) continue;
    if (str1Dic[curWord] === undefined) str1Dic[curWord] = 1;
    else str1Dic[curWord] += 1;
  }

  for (let i = 0; i < str2.length - 1; i++) {
    let curWord = str2[i] + str2[i + 1];
    curWord = curWord.match(/[a-zA-Z]+/g);
    if (curWord === null || curWord[0].length === 1) continue;
    if (str2Dic[curWord] === undefined) str2Dic[curWord] = 1;
    else str2Dic[curWord] += 1;
  }
  // 2. 교집합 차집합 구하기
  let 교집합 = 0;
  let 합집합 = 0;
  // 교집합
  for (let key in str1Dic) {
    if (str1Dic[key] !== undefined && str2Dic[key] !== undefined)
      교집합 += Math.min(str2Dic[key], str1Dic[key]);
  }
  // 합집합
  for (let key in str1Dic) {
    if (str2Dic[key] !== undefined && str2Dic[key] < str1Dic[key])
      str2Dic[key] = str1Dic[key];
    else if (str2Dic[key] === undefined) str2Dic[key] = str1Dic[key];
  }
  for (let key in str2Dic) {
    합집합 += str2Dic[key];
  }
  // 합집합(분모)이 0일때의 예외처리
  if (합집합 === 0) return 65536;
  let 자카드유사도 = 교집합 / 합집합;
  return Math.floor(자카드유사도 * 65536);
}
